src/deploy/functions/release.ts

```diff
@@ -0,0 +1,477 @@
+  bindings: [
+    {
+      role: "roles/cloudfunctions.invoker",
+      members: ["allUsers"],
+    },
+  ],
+};
+
+function startTimer(name: string, type: string) {
+  timings[name] = { type: type, t0: process.hrtime() };
+}
+
+function endTimer(name: string) {
+  if (!timings[name]) {
+    logger.debug("[functions] no timer initialized for", name);
+    return;
+  }
+
+  // hrtime returns a duration as an array of [seconds, nanos]
+  const duration = process.hrtime(timings[name].t0);
+  track(
+    "Functions Deploy (Duration)",
+    timings[name].type,
+    duration[0] * 1000 + Math.round(duration[1] * 1e-6)
+  );
+}
+
+async function fetchTriggerUrls(projectId: string, ops: helper.Operation[], sourceUrl: string) {
+  if (!_.find(ops, ["trigger.httpsTrigger", {}])) {
+    // No HTTPS functions being deployed
+    return;
+  }
+
+  const functions = await cloudfunctions.listAllFunctions(projectId);
+  const httpFunctions = _.chain(functions)
+    .filter({ sourceUploadUrl: sourceUrl })
+    .filter("httpsTrigger")
+    .value();
+  _.forEach(httpFunctions, (httpFunc) => {
+    const op = _.find(ops, { funcName: httpFunc.name });
+    if (op) {
+      op.triggerUrl = httpFunc.httpsTrigger.url;
+    }
+  });
+  return;
+}
+
+function printSuccess(op: helper.Operation) {
+  endTimer(op.funcName);
+  utils.logSuccess(
+    clc.bold.green("functions[" + helper.getFunctionLabel(op.funcName) + "]: ") +
+      "Successful " +
+      op.type +
+      " operation. "
+  );
+  if (op.triggerUrl && op.type !== "delete") {
+    logger.info(
+      clc.bold("Function URL"),
+      "(" + helper.getFunctionName(op.funcName) + "):",
+      op.triggerUrl
+    );
+  }
+}
+
+function printFail(op: helper.Operation) {
+  endTimer(op.funcName);
+  failedDeployments.push(helper.getFunctionName(op.funcName));
+  utils.logWarning(
+    clc.bold.yellow("functions[" + helper.getFunctionLabel(op.funcName) + "]: ") +
+      "Deployment error."
+  );
+  if (op.error?.code === 8) {
+    logger.debug(op.error.message);
+    logger.info(
+      "You have exceeded your deployment quota, please deploy your functions in batches by using the --only flag, " +
+        "and wait a few minutes before deploying again. Go to " +
+        clc.underline("https://firebase.google.com/docs/cli/#deploy_specific_functions") +
+        " to learn more."
+    );
+  } else {
+    logger.info(op.error?.message);
+  }
+}
+
+function printTooManyOps(projectId: string) {
+  utils.logWarning(
+    clc.bold.yellow("functions:") + " too many functions are being deployed, cannot poll status."
+  );
+  logger.info(
+    "In a few minutes, you can check status at " + utils.consoleUrl(projectId, "/functions/logs")
+  );
+  logger.info(
+    "You can use the --only flag to deploy only a portion of your functions in the future."
+  );
+  deployments = []; // prevents analytics tracking of deployments
+}
+
+export async function release(context: any, options: any, payload: any): Promise<void> {
```

Comment:
This is a 340LOC method that has nested conditionals since we can't early exit anywhere. It would probably be cleaner to actually use helper methods. Just a drive-by comment since this code was just translated, not written in the PR.

Reply:
Agree 100 percent - this is getting torn to shreds in a coming PR